lib/puppet/configurer.rb

```diff
@@ -109,8 +109,17 @@ def convert_catalog(result, duration, options = {})
 
     catalog_conversion_time = thinmark do
       # Will mutate the result and replace all Deferred values with resolved values
-      if options[:convert_for_node] && options[:convert_with_facts]
-        Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(options[:convert_for_node], options[:convert_with_facts], result)
+      if facts = options[:convert_with_facts]
+        Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(facts, result)
+      else
+        # During a cached_catalog run, facts may not have been collected; even in
+        # that case, we should still attempt to reolve_and_replace deferred
+        # functions and send an empty hash for facts parameter.
+        #
+        # Perhaps the prepare_and_retrieve_catalog_from_cache method
+        # could check a Puppet setting to determine if it should run facter or
+        # not?
+        Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace({}, result)
```

Comment:
I'm not clear why facts may not have been collected with a cached catalog run. They're still used by providers.

Reply:
If I run `puppet agent -t --use_cached_catalog`, the `convert_catalog` is called before `get_facts` is. Given Henrik's comments, we might just want to ensure that facts are always run earlier so that the user cannot apply a catalog without facts.